chunkstorage.go

```diff
@@ -3,71 +3,47 @@ package desync
 import (
 	"bytes"
 	"context"
-	"crypto/sha512"
 	"fmt"
 	"os"
 	"sync"
 
-	"io"
-
 	"github.com/pkg/errors"
 )
 
-type ChunkJob struct {
-	num   int
-	chunk IndexChunk
-}
-
 type ChunkStorage struct {
 	sync.Mutex
-	ctx      context.Context
-	cancel   context.CancelFunc
-	chopFile string
-	n        int
-	ws       WriteStore
-	in       <-chan ChunkJob
-	pb       ProgressBar
-
-	wg      sync.WaitGroup
-	results map[int]IndexChunk
-	pErr    error
-}
-
-// Stores the chinks passed in the input channel. If chopFile is passed (non-blank string), then the chunks will be
-// read from that file. Otherwise it's assumed that the chunk contents are pre-loaded and passed via the input channel.
-func NewChunkStorage(ctx context.Context, cancel context.CancelFunc, chopFile string, n int, ws WriteStore, in <-chan ChunkJob, pb ProgressBar) *ChunkStorage {
-	return &ChunkStorage{
-		ctx:      ctx,
-		cancel:   cancel,
-		n:        n,
-		ws:       ws,
-		chopFile: chopFile,
-		in:       in,
-		pb:       pb,
-		wg:       sync.WaitGroup{},
-		results:  make(map[int]IndexChunk),
-	}
+	ctx    context.Context
+	cancel context.CancelFunc
+	n      int
+	ws     WriteStore
+	in     <-chan Chunk
+	pb     ProgressBar
+
+	wg     sync.WaitGroup
+	pErr   error
+	stored map[ChunkID]bool
 }
 
-func readChunkFromFile(f *os.File, c IndexChunk) ([]byte, error) {
-	var err error
-	b := make([]byte, c.Size)
-
-	// Position the filehandle to the place where the chunk is meant to come
-	// from within the file
-	if _, err = f.Seek(int64(c.Start), io.SeekStart); err != nil {
-		return b, err
+// Stores chunks passed in the input channel asynchronously. Wait() will wait for until the input channel is closed or
+// until there's an error, in which case it will return it.
+func NewChunkStorage(ctx context.Context, cancel context.CancelFunc, n int, ws WriteStore, in <-chan Chunk, pb ProgressBar) *ChunkStorage {
+	s := &ChunkStorage{
+		ctx:    ctx,
+		cancel: cancel,
+		n:      n,
+		ws:     ws,
+		in:     in,
+		pb:     pb,
+		stored: make(map[ChunkID]bool),
```

Comment:
I believe it's a tiny bit more efficient to use a map[ChunkID]struct{} here. Not a big deal though

Reply:
Zero bytes!, I'll take it :)